builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,380 @@
+				if err := framework.DeleteWAL(ctx, s, walEntry.walID); err != nil {
+					log.Warn("unable to delete WAL", "error", err, "WAL ID", walEntry.walID)
+				}
+			} else {
+				log.Info("adjusting priority for Role")
+				item.Value = walEntry.walID
+				item.Priority = time.Now().Unix()
+			}
+		}
+
+		if err := b.pushItem(&item); err != nil {
+			log.Warn("unable to enqueue item", "error", err, "role", roleName)
+		}
+	}
+}
+
+// runTicker kicks off a periodic ticker that invoke the automatic credential
+// rotation method at a determined interval. The default interval is 5 seconds.
+func (b *databaseBackend) runTicker(ctx context.Context, s logical.Storage) {
+	b.logger.Info("starting periodic ticker")
+	tick := time.NewTicker(queueTickInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			b.rotateCredentials(ctx, s)
+
+		case <-ctx.Done():
+			b.logger.Info("stopping periodic ticker")
+			return
+		}
+	}
+}
+
+// setCredentialsWAL is used to store information in a WAL that can retry a
+// credential setting or rotation in the event of partial failure.
+type setCredentialsWAL struct {
+	NewPassword string
+	OldPassword string
+	RoleName    string
+	Username    string
+
+	LastVaultRotation time.Time
+
+	walID string
+}
+
+// rotateCredentials sets a new password for a static account. This method is
+// invoked in the runTicker method, which is in it's own go-routine, and invoked
+// periodically (approximately every 5 seconds).
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		// quit rotating credentials if shutdown has started
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		item, err := b.popItem()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// guard against possible nil item
+		if item == nil {
+			return nil
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
+		}
+		if role == nil {
+			b.logger.Warn("role not found", "role", item.Key, "error", err)
+			continue
+		}
+
+		// if "now" is less than the Item priority, then this item does not need to
+		// be rotated
+		if time.Now().Unix() < item.Priority {
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			// break out of the for loop
+			break
+		}
+
+		input := &setStaticAccountInput{
+			RoleName: item.Key,
+			Role:     role,
+		}
+
+		// If there is a WAL entry related to this Role, the corresponding WAL ID
+		// should be stored in the Item's Value field.
+		if walID, ok := item.Value.(string); ok {
+			walEntry := b.findStaticWAL(ctx, s, walID)
```

Comment:
I think we probably want to check for a storage error here and push the value back onto the queue with a backoff.

Reply:
Is there a type or group of storage errors we would want to check?

Reply:
added in 46c9a1d2e